### ISSUE TITLE
apt fails on fips-enabled hosts

### ISSUE THREAD

**Issue by mkim37** [jenkinsci/docker#1694]
### Jenkins and plugins versions report
Jenkins LTS 2.414.1

### What Operating System are you using (both controller, and any agents involved in the problem)?

Debian 12 Bookworm

### Reproduction steps

```
[root@jumpbox ~]# docker run --rm -it -u root jenkins/jenkins:2.414.1-jdk11 bash
root@7bf45452930d:/# cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
NAME="Debian GNU/Linux"
VERSION_ID="12"
VERSION="12 (bookworm)"
VERSION_CODENAME=bookworm
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
root@7bf45452930d:/# sysctl crypto.fips_enabled
crypto.fips_enabled = 1
root@7bf45452930d:/# apt-get update
Get:1 http://deb.debian.org/debian bookworm InRelease [151 kB]
Get:2 http://deb.debian.org/debian bookworm-updates InRelease [52.1 kB]
Get:3 http://deb.debian.org/debian-security bookworm-security InRelease [48.0 kB]
Get:4 http://deb.debian.org/debian bookworm/main amd64 Packages [8906 kB]
Get:6 http://deb.debian.org/debian bookworm-updates/main amd64 Packages [4952 B]
Get:7 http://deb.debian.org/debian-security bookworm-security/main amd64 Packages [58.3 kB]
Get:5 https://packagecloud.io/github/git-lfs/debian bookworm InRelease [25.8 kB]
Get:8 https://packagecloud.io/github/git-lfs/debian bookworm/main amd64 Packages [2131 B]
Fetched 9248 kB in 2s (5902 kB/s)
fatal error in libgcrypt, file ../../src/misc.c, line 92, function _gcry_fatal_error: requested algo not in md context

Fatal error: requested algo not in md context
Aborted

```

### Expected Results

No error

### Actual Results

Error

### Anything else?

https://github.com/jenkinsci/docker/issues/1686
https://github.com/debuerreotype/docker-debian-artifacts/issues/202

**Comment by dduportal** [comment]
Hi @mkim37 thanks for reporting.

Debian operating system is not intended to operate in FIPS as per https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1014517#37: this error with `apt` might never go away.

I don't see a solution to this behavior: as far as I can tell, the Jenkins project never worked on FIPS support.

You might try another declination of the Jenkins controller image such as Alpine but I have no idea if it works in FIPS as well.
You might also have to run Jenkins itself outside a container or build your own container image of Jenkins. 

For info, other users tried to run Jenkins in FIPS environments: https://community.jenkins.io/t/jenkins-fails-on-fips-mode-enabled-rhel/2714/16 but it highly depends on the plugins you are using.


This is not a bug in Jenkins neither its Docker image: as per I'm closing this issue. Feel free to raise the FIPS topic in the [community forum](https://community.jenkins.io) or join the https://www.jenkins.io/sigs/platform/ if you want to bring this topic on the table and invest time in driving it.

**Comment by dduportal** [comment]
@mkim37 additionally, you can also try with CloudBees: https://www.cloudbees.com/newsroom/cloudbees-releases-hardened-cloudbees-ci-enable-secure-dod-compliant-software-delivery

**Comment by mkim37** [comment]
@dduportal Thanks for replying. This was more of an FYI than a bug report. It appears to have been fixed in debian:bookworm-20230725

**Comment by dduportal** [comment]
> @dduportal Thanks for replying. This was more of an FYI than a bug report.

Thanks for hte FYI @mkim37 !

> It appears to have been fixed in debian:bookworm-20230725

The `jenkins/jenkins:2.414.1-jdk11` image is built on top on `debian:bookworm-20230814`:
https://github.com/jenkinsci/docker/blob/e1c29ed1273932d3f118939463440f0ea97fb49f/11/debian/bookworm/hotspot/Dockerfile#L2

Did you mean `debian:bookworm-20230825` ?

**Comment by mkim37** [comment]
@dduportal Sorry I just saw your reply. It's working in debian:bookworm-20230725 and presumably after but not before. I haven't had the chance to look through the debian bug tracker to see what changed and why. Strangely it's still not working in the LTS image.

```
# https://hub.docker.com/_/debian/tags?page=1&name=bookworm-202306
[root@mcaas-jumpbox ~]# docker run --rm -it debian:bookworm-20230612 apt-get update
Unable to find image 'debian:bookworm-20230612' locally
bookworm-20230612: Pulling from library/debian
bba7bb10d5ba: Pull complete
Digest: sha256:d568e251e460295a8743e9d5ef7de673c5a8f9027db11f4e666e96fb5bed708e
Status: Downloaded newer image for debian:bookworm-20230612
Get:1 http://deb.debian.org/debian bookworm InRelease [151 kB]
Get:2 http://deb.debian.org/debian bookworm-updates InRelease [52.1 kB]
Get:3 http://deb.debian.org/debian-security bookworm-security InRelease [48.0 kB]
Get:4 http://deb.debian.org/debian bookworm/main amd64 Packages [8906 kB]
Get:5 http://deb.debian.org/debian bookworm-updates/main amd64 Packages [4952 B]
Get:6 http://deb.debian.org/debian-security bookworm-security/main amd64 Packages [58.3 kB]
Fetched 9220 kB in 2s (5814 kB/s)
fatal error in libgcrypt, file ../../src/misc.c, line 92, function _gcry_fatal_error: requested algo not in md context

Fatal error: requested algo not in md context

# https://hub.docker.com/_/debian/tags?page=1&name=bookworm-202307
[root@jumpbox ~]# docker run --rm -it debian:bookworm-20230703 apt-get update
Unable to find image 'debian:bookworm-20230703' locally
bookworm-20230703: Pulling from library/debian
d52e4f012db1: Pull complete
Digest: sha256:3d868b5eb908155f3784317b3dda2941df87bbbbaa4608f84881de66d9bb297b
Status: Downloaded newer image for debian:bookworm-20230703
Get:1 http://deb.debian.org/debian bookworm InRelease [151 kB]
Get:2 http://deb.debian.org/debian bookworm-updates InRelease [52.1 kB]
Get:3 http://deb.debian.org/debian-security bookworm-security InRelease [48.0 kB]
Get:4 http://deb.debian.org/debian bookworm/main amd64 Packages [8906 kB]
Get:5 http://deb.debian.org/debian bookworm-updates/main amd64 Packages [4952 B]
Get:6 http://deb.debian.org/debian-security bookworm-security/main amd64 Packages [58.3 kB]
Fetched 9220 kB in 2s (4216 kB/s)
fatal error in libgcrypt, file ../../src/misc.c, line 92, function _gcry_fatal_error: requested algo not in md context

Fatal error: requested algo not in md context

[root@jumpbox ~]# docker run --rm -it debian:bookworm-20230725 apt-get update
Unable to find image 'debian:bookworm-20230725' locally
bookworm-20230725: Pulling from library/debian
785ef8b9b236: Pull complete
Digest: sha256:9f76a008888da28c6490bedf7bdaa919bac9b2be827afd58d6eb1b916e1e5918
Status: Downloaded newer image for debian:bookworm-20230725
Get:1 http://deb.debian.org/debian bookworm InRelease [151 kB]
Get:2 http://deb.debian.org/debian bookworm-updates InRelease [52.1 kB]
Get:3 http://deb.debian.org/debian-security bookworm-security InRelease [48.0 kB]
Get:4 http://deb.debian.org/debian bookworm/main amd64 Packages [8906 kB]
Get:5 http://deb.debian.org/debian bookworm-updates/main amd64 Packages [4952 B]
Get:6 http://deb.debian.org/debian-security bookworm-security/main amd64 Packages [58.3 kB]
Fetched 9220 kB in 2s (5454 kB/s)
Reading package lists... Done

[root@jumpbox ~]# docker run --rm -it debian:bookworm-20230814 apt-get update
Unable to find image 'debian:bookworm-20230814' locally
bookworm-20230814: Pulling from library/debian
Digest: sha256:b91baba9c2cae5edbe3b0ff50ae8f05157e3ae6f018372dcfc3aba224acb392b
Status: Downloaded newer image for debian:bookworm-20230814
Get:1 http://deb.debian.org/debian bookworm InRelease [151 kB]
Get:2 http://deb.debian.org/debian bookworm-updates InRelease [52.1 kB]
Get:3 http://deb.debian.org/debian-security bookworm-security InRelease [48.0 kB]
Get:4 http://deb.debian.org/debian bookworm/main amd64 Packages [8906 kB]
Get:5 http://deb.debian.org/debian bookworm-updates/main amd64 Packages [4952 B]
Get:6 http://deb.debian.org/debian-security bookworm-security/main amd64 Packages [58.3 kB]
Fetched 9220 kB in 3s (3193 kB/s)
Reading package lists... Done

[root@jumpbox ~]# docker run --rm -it -u root jenkins/jenkins:2.414.1-jdk11 apt-get update
Unable to find image 'jenkins/jenkins:2.414.1-jdk11' locally
2.414.1-jdk11: Pulling from jenkins/jenkins
de4cac68b616: Already exists
8252e4beaf27: Pull complete
88f68f285a52: Pull complete
1c66d3c5c634: Pull complete
554bfe5cc2ad: Pull complete
921190312de1: Pull complete
4bf15d2379ee: Pull complete
de6ad7a9d923: Pull complete
c591934e9a70: Pull complete
e3c5028ef70a: Pull complete
a95164b0df40: Pull complete
cbd7a2671ac8: Pull complete
66ceb70ec890: Pull complete
Digest: sha256:645d55b80126b67dd23b6ba65d455ebc13689d057f498fbecd7bdb9165f0dd64
Status: Downloaded newer image for jenkins/jenkins:2.414.1-jdk11
Get:1 http://deb.debian.org/debian bookworm InRelease [151 kB]
Get:2 http://deb.debian.org/debian bookworm-updates InRelease [52.1 kB]
Get:3 http://deb.debian.org/debian-security bookworm-security InRelease [48.0 kB]
Get:4 http://deb.debian.org/debian bookworm/main amd64 Packages [8906 kB]
Get:6 http://deb.debian.org/debian bookworm-updates/main amd64 Packages [4952 B]
Get:5 https://packagecloud.io/github/git-lfs/debian bookworm InRelease [25.8 kB]
Get:7 http://deb.debian.org/debian-security bookworm-security/main amd64 Packages [58.3 kB]
Get:8 https://packagecloud.io/github/git-lfs/debian bookworm/main amd64 Packages [2131 B]
Fetched 9248 kB in 2s (5408 kB/s)
fatal error in libgcrypt, file ../../src/misc.c, line 92, function _gcry_fatal_error: requested algo not in md context

Fatal error: requested algo not in md context
```

**Comment by mkim37** [comment]
@dduportal This is being caused by https://github.com/jenkinsci/docker/blob/master/11/debian/bookworm/hotspot/Dockerfile#L30

Assuming this won't be fixed for 2.414.2, we'll have to switch to the rhel-ubi8 image.

```
#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 4.57kB done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/library/debian:bookworm-20230904
#3 DONE 0.0s

#4 [internal] load metadata for docker.io/library/eclipse-temurin:11.0.20_8-jdk-jammy
#4 DONE 0.3s

#5 [stage-1  1/13] FROM docker.io/library/debian:bookworm-20230904
#5 DONE 0.0s

#6 [jre-build 1/2] FROM docker.io/library/eclipse-temurin:11.0.20_8-jdk-jammy@sha256:f215a785cafb0b58474de3dacdb28dde19c3f89383de070ad0362f40ddf0d2bb
#6 DONE 0.0s

#7 [stage-1  2/13] RUN apt-get update   && apt-get install -y --no-install-recommends     ca-certificates     curl     git     gnupg     gpg     libfontconfig1     libfreetype6     procps     ssh-client     tini     unzip     tzdata   && rm -rf /var/lib/apt/lists/*
#7 CACHED

#8 [jre-build 2/2] RUN jlink          --add-modules ALL-MODULE-PATH          --no-man-pages          --compress=2          --output /javaruntime
#8 CACHED

#9 [internal] load build context
#9 transferring context: 357B done
#9 DONE 0.0s

#10 [stage-1  3/13] RUN curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh -o /tmp/script.deb.sh   && bash /tmp/script.deb.sh   && rm -f /tmp/script.deb.sh   && apt-get install -y --no-install-recommends     git-lfs   && rm -rf /var/lib/apt/lists/*   && git lfs install
#10 0.864 Detected operating system as debian/12.
#10 0.864 Checking for curl...
#10 0.864 Detected curl...
#10 0.864 Checking for gpg...
#10 0.864 Detected gpg...
#10 0.880 Detected apt version as 2.6.1
#10 0.880 Running apt-get update... done.
#10 3.029 Installing debian-archive-keyring which is needed for installing
#10 3.029 apt-transport-https on many Debian systems.
#10 3.883 Installing apt-transport-https... done.
#10 5.110 Installing /etc/apt/sources.list.d/github_git-lfs.list...done.
#10 5.464 Importing packagecloud gpg key... gpg: out of core handler ignored in FIPS mode
#10 5.936 Packagecloud gpg key imported to /etc/apt/keyrings/github_git-lfs-archive-keyring.gpg
#10 5.936 done.
#10 5.936 Running apt-get update... /tmp/script.deb.sh: line 119:   168 Aborted                 apt-get update &> /dev/null
#10 7.664 done.
#10 7.664
#10 7.664 The repository is setup! You can now install packages.
#10 7.675 Reading package lists...fatal error in libgcrypt, file ../../src/misc.c, line 92, function _gcry_fatal_error: requested algo not in md context
#10 8.242
#10 8.242 Fatal error: requested algo not in md context
#10 8.277 Aborted (core dumped)
#10 ERROR: process "/bin/sh -c curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh -o /tmp/script.deb.sh   && bash /tmp/script.deb.sh   && rm -f /tmp/script.deb.sh   && apt-get install -y --no-install-recommends     git-lfs   && rm -rf /var/lib/apt/lists/*   && git lfs install" did not complete successfully: exit code: 134
------
 > [stage-1  3/13] RUN curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh -o /tmp/script.deb.sh   && bash /tmp/script.deb.sh   && rm -f /tmp/script.deb.sh   && apt-get install -y --no-install-recommends     git-lfs   && rm -rf /var/lib/apt/lists/*   && git lfs install:
#10 0.864 Detected gpg...
done.
#10 3.029 Installing debian-archive-keyring which is needed for installing
#10 3.029 apt-transport-https on many Debian systems.
done.
#10 5.464 Importing packagecloud gpg key... gpg: out of core handler ignored in FIPS mode
fatal error in libgcrypt, file ../../src/misc.c, line 92, function _gcry_fatal_error: requested algo not in md context
#10 8.242
#10 8.242 Fatal error: requested algo not in md context
#10 8.277 Aborted (core dumped)
------
Dockerfile:29
--------------------
  28 |
  29 | >>> RUN curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh -o /tmp/script.deb.sh \
  30 | >>>   && bash /tmp/script.deb.sh \
  31 | >>>   && rm -f /tmp/script.deb.sh \
  32 | >>>   && apt-get install -y --no-install-recommends \
  33 | >>>     git-lfs \
  34 | >>>   && rm -rf /var/lib/apt/lists/* \
  35 | >>>   && git lfs install
  36 |
--------------------
ERROR: failed to solve: process "/bin/sh -c curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh -o /tmp/script.deb.sh   && bash /tmp/script.deb.sh   && rm -f /tmp/script.deb.sh   && apt-get install -y --no-install-recommends     git-lfs   && rm -rf /var/lib/apt/lists/*   && git lfs install" did not complete successfully: exit code: 134
```
